File: graph/model/sector.go
package model

import (
	"database/sql"

	"github.com/strahe/curio-dashboard/types"
)

type Sector struct {
	SpID      types.Address
	SectorNum int
	Meta      *SectorMeta `json:"meta,omitempty"`
}

type SectorMeta struct {
	SpID            types.Address    `json:"spId"`
	SectorNum       int              `json:"sectorNum"`
	RegSealProof    int              `json:"regSealProof"`
	TicketEpoch     int              `json:"ticketEpoch"`
	TicketValue     types.ByteArray  `json:"ticketValue,omitempty"`
	OrigSealedCid   string           `json:"origSealedCid"`
	OrigUnsealedCid string           `json:"origUnsealedCid"`
	CurSealedCid    string           `json:"curSealedCid"`
	CurUnsealedCid  string           `json:"curUnsealedCid"`
	MsgCidPrecommit sql.Null[string] `json:"msgCidPrecommit,omitempty"`
	MsgCidCommit    sql.Null[string] `json:"msgCidCommit,omitempty"`
	MsgCidUpdate    sql.Null[string] `json:"msgCidUpdate,omitempty"`
	SeedEpoch       int              `json:"seedEpoch"`
	SeedValue       types.ByteArray  `json:"seedValue,omitempty"`
	ExpirationEpoch sql.Null[int]    `json:"expirationEpoch,omitempty"`
	IsCC            bool             `json:"isCC"`
	Deadline        sql.Null[int]    `json:"deadline,omitempty"`
	Partition       sql.Null[int]    `json:"partition,omitempty"`
}

type SectorLocation struct {
	MinerID        types.Address    `json:"minerId"`
	SectorNum      int              `json:"sectorNum"`
	SectorFiletype int              `json:"sectorFiletype"`
	StorageID      string           `json:"storageId"`
	IsPrimary      sql.Null[bool]   `json:"isPrimary,omitempty"`
	ReadTS         sql.Null[string] `json:"readTs,omitempty"`
	ReadRefs       int              `json:"readRefs"`
	WriteTS        sql.Null[string] `json:"writeTs,omitempty"`
	WriteLockOwner sql.Null[string] `json:"writeLockOwner,omitempty"`
}

File: ui/src/views/query/sector.ts
import gql from 'graphql-tag'
import { taskFragment, taskHistoryFragment } from '@/views/query/task'
import { porepAllFragment } from '@/views/query/porep'

const metaFragment = gql`
  fragment MetaAll on SectorMeta {
    id
    spId
    sectorNum
    regSealProof
    ticketEpoch
    ticketValue
    origSealedCid
    origUnsealedCid
    curSealedCid
    curUnsealedCid
    msgCidPrecommit
    msgCidCommit
    msgCidUpdate
    seedEpoch
    seedValue
    expirationEpoch
    isCC
  }
`

const locationFragment = gql`
  fragment LocationAll on SectorLocation {
    minerId
    sectorNum
    sectorFiletype
    storageId
    isPrimary
    readTs
    readRefs
    writeTs
    writeLockOwner
  }
`

export const GetSectors = gql`
  query GetSectors($miner: ActorID, $sectorNumber: Int, $offset: Int!, $limit: Int!) {
    sectors(actor: $miner, sectorNumber: $sectorNumber, offset: $offset, limit: $limit) {
      id
      meta {
        ...MetaAll
      }
      locations {
        ...LocationAll
      }
    }
    sectorsCount(actor: $miner)
    actors {
      address
    }
  }
  ${metaFragment}
  ${locationFragment}
`

export const GetSectorMeta = gql`
  query GetSectorMeta($miner: ActorID!, $sectorNumber: Int!) {
    sector(actor: $miner, sectorNumber: $sectorNumber) {
      id
      status
      meta {
        ...MetaAll
      }
    }
  }
  ${metaFragment}
`

export const GetSectorPoRep = gql`
  query GetSectorPoRep($miner: ActorID!, $sectorNumber: Int!) {
    sector(actor: $miner, sectorNumber: $sectorNumber) {
      id
      porep {
        ...PorepAll
      }
    }
  }
  ${porepAllFragment}
`

export const GetSectorPieces = gql`
  query GetSectorPieces($miner: ActorID!, $sectorNumber: Int!) {
    sector(actor: $miner, sectorNumber: $sectorNumber) {
      id
      pieces {
        spID
        sectorNum
        pieceNum
        pieceCID
        pieceSize
        requestedKeepData
        rawDataSize
        startEpoch
        origEndEpoch
        f05DealID
        ddoPam
        f05DealProposal
      }
    }
  }
`

export const GetSectorLocations = gql`
  query GetSectorLocations($miner: ActorID!, $sectorNumber: Int!) {
    sector(actor: $miner, sectorNumber: $sectorNumber) {
      id
      locations {
        ...LocationAll
      }
    }
  }
  ${locationFragment}
`

export const GetSectorEvents = gql`
  query GetSectorEvents($miner: ActorID!, $sectorNumber: Int!) {
    sector(actor: $miner, sectorNumber: $sectorNumber) {
      id
      events {
        ...TaskHistoryAll
      }
    }
  }
  ${taskHistoryFragment}
`

export const GetSectorTasks = gql`
    query GetSectorTasks($miner: ActorID!, $sectorNumber: Int!) {
      sector(actor: $miner, sectorNumber: $sectorNumber) {
        id
        spID
        sectorNum
        tasks {
          ...TaskAll
        }
      }
    }
    ${taskFragment}
`

export const RemoveSector = gql`
    mutation RemoveSector($miner: ActorID!, $sectorNumber: Int!) {
      removeSector(miner: $miner, sectorNumber: $sectorNumber)
    }
`
